dpppt-backend-sdk/dpppt-backend-sdk-ws/src/test/java/org/dpppt/backend/sdk/ws/controller/GaenControllerTest.java

```diff
@@ -1354,6 +1388,64 @@ public void testMalciousTokenFails() throws Exception {
     assertTrue(authenticateError.contains("Unsigned Claims JWTs are not supported."));
   }
 
+  @Test
+  public void testUploadTodaysKeyWillBeReleasedTomorrow() throws Exception {
+    GaenRequest exposeeRequest = new GaenRequest();
+    List<GaenKey> keys = new ArrayList<>();
+    for (int i = 0; i < 30; i++) {
+      var tmpKey = new GaenKey();
+      tmpKey.setRollingStartNumber((int) UTCInstant.today().minusDays(i).get10MinutesSince1970());
+      var keyData = String.format("testKey32Bytes%02d", i);
+      tmpKey.setKeyData(Base64.getEncoder().encodeToString(keyData.getBytes("UTF-8")));
+      tmpKey.setRollingPeriod(144);
+      tmpKey.setFake(0);
+      tmpKey.setTransmissionRiskLevel(0);
+      keys.add(tmpKey);
+    }
+    exposeeRequest.setGaenKeys(keys);
+
+    var delayedKeyDateSent = (int) UTCInstant.today().get10MinutesSince1970();
+    exposeeRequest.setDelayedKeyDate(delayedKeyDateSent);
+
+    String token = createToken(UTCInstant.now().plusMinutes(5));
+    MvcResult responseAsync =
+        mockMvc
+            .perform(
+                post("/v1/gaen/exposed")
+                    .contentType(MediaType.APPLICATION_JSON)
+                    .header("Authorization", "Bearer " + token)
+                    .header("User-Agent", androidUserAgent)
+                    .content(json(exposeeRequest)))
+            .andExpect(request().asyncStarted())
+            .andReturn();
+    mockMvc.perform(asyncDispatch(responseAsync)).andExpect(status().isOk());
+
+    var tooEarlyInstant = UTCInstant.today();
+    MockHttpServletResponse response =
+        mockMvc
+            .perform(
+                get("/v1/gaen/exposed/" + tooEarlyInstant.getTimestamp())
+                    .header("User-Agent", androidUserAgent))
+            .andExpect(status().is(204))
+            .andReturn()
+            .getResponse();
+
+    Clock fourAMTomorrow =
+        Clock.fixed(UTCInstant.today().plusDays(1).plusHours(4).getInstant(), ZoneOffset.UTC);
+
+    UTCInstant.setClock(fourAMTomorrow);
```

Comment:
To avoid this, could you pass the `now` to `mockMvc`?

Reply:
After a bit of research I think our only option would be to do something like `https://github.com/raphw/byte-buddy`. I tried the `Mockito` Framework, but that one only allows to replace a class with _all_ stubs and redefine functions. I think this is not really feasible, since it would blow up the test code, since we would have to redefine all functions of the mocked class.

Also the `PowerMockito` plugin seems to be rather verbose, when our only intention is to manipulate actual system time.

Another option would be to replace the static `UTCInstant` approach with a singleton approach using `Beans` and dependency injection. I'm not sure how easy it is to change everything and inject everywhere the `Bean` (maybe with `@Autowired` but then we introduce more magic...)

A way to prevent forgetting the `resetClock` could be to implement the `Disposable` interface and use it with a `try-with` statement.

Some remarks on test parallelisation: 
> Since every test class itself initialises a new VM instance (and hence a new static object is instantiated) as far as I understand it, and tests aren't parallelised within test classes (this would anyways be catastrophic, since for the integration tests we relay on database inserts and then we would have to mock them as well), this should not be problematic. The only "dangerous" thing is that one could use the `setClock` function outside test-packages, which is currently not prevented (though we could check for `org.junit` being on the classPath).

Maybe I'm also completely on the wrong path here, but in my opinion it is easier that way. 

@martinalig any input? :)